cpp/src/arrow/ipc/reader.cc

```diff
@@ -89,16 +90,96 @@ Status InvalidMessageType(Message::Type expected, Message::Type actual) {
     }                                                                   \
   } while (0)
 
-}  // namespace
-
 // ----------------------------------------------------------------------
 // Record batch read path
 
-/// Accessor class for flatbuffers metadata
-class IpcComponentSource {
+/// The field_index and buffer_index are incremented based on how much of the
+/// batch is "consumed" (through nested data reconstruction, for example)
+class ArrayLoader {
  public:
-  IpcComponentSource(const flatbuf::RecordBatch* metadata, io::RandomAccessFile* file)
-      : metadata_(metadata), file_(file) {}
+  explicit ArrayLoader(const flatbuf::RecordBatch* metadata, io::RandomAccessFile* file,
+                       const DictionaryMemo* dictionary_memo,
+                       const IpcReadOptions& options,
+                       Compression::type compression = Compression::UNCOMPRESSED)
+      : metadata_(metadata),
+        file_(file),
+        dictionary_memo_(dictionary_memo),
+        options_(options),
+        compression_(compression),
+        max_recursion_depth_(options.max_recursion_depth),
+        buffer_index_(0),
+        field_index_(0),
+        skip_io_(false) {}
+
+  Status ReadBuffer(int64_t offset, int64_t length, std::shared_ptr<Buffer>* out) {
+    if (skip_io_) {
+      return Status::OK();
+    }
+
+    // This construct permits overriding GetBuffer at compile time
+    if (!BitUtil::IsMultipleOf8(offset)) {
+      return Status::Invalid("Buffer ", buffer_index_,
+                             " did not start on 8-byte aligned offset: ", offset);
+    }
+    return file_->ReadAt(offset, length).Value(out);
+  }
+
+  // Use this to disable calls to RandomAccessFile::ReadAt, for field skipping
+  void SkipIO(bool skip_io = true) { skip_io_ = skip_io; }
+
+  Status LoadType(const DataType& type) { return VisitTypeInline(type, this); }
+
+  Status DecompressBuffers() {
+    // If the buffers are indicated to be compressed, instantiate the codec and
+    // decompress them
+    std::unique_ptr<util::Codec> codec;
+    ARROW_ASSIGN_OR_RAISE(codec, util::Codec::Create(compression_));
+
+    // TODO: Parallelize decompression
```

Comment:
Right. The parallelization needs to happen at the field-level, which will require some refactoring. I'll open a follow up JIRA since it's out of scope for this PR